### PR TITLE
Change patrol output

### PR DESCRIPTION
**Pull request by pdenert**
This PR completely changes the current output when firing patrol tests. It also adds two new flags:
- showTestSteps,
- showFlutterLogs.

To reach final effect, we need to handle exceptions during test, but this require #2362.
It's also adding new package: patrol_log, that need setup of the pubspec, changelog etc.

Preview of the new output on the video:

https://github.com/user-attachments/assets/518f151a-b0a6-4fb7-87e7-ca9e5c86842f

### CODE REVIEW

File: packages/patrol_log/lib/src/patrol_log_reader.dart [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+              !_skippedTests.contains(testEntry.name)) {
+            _skippedTests.add(testEntry.name);
+            _controller.add(entry);
+          } else if (testEntry.status != TestEntryStatus.skip) {
+            _controller.add(entry);
+          }
+        } else {
+          _controller.add(entry);
+        }
+      }
```
Comment (piotruela) [inline packages/patrol_log/lib/src/patrol_log_reader.dart:92]: I would move this block of code to separate private method with a little explanation what is happening (and especially a name explaining what's happening inside)

File: packages/patrol_log/lib/src/patrol_log_reader.dart [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    } else if (showFlutterLogs && line.contains('Runner: (Flutter) flutter:')) {
+      final regExp = RegExp(r'Runner: \(Flutter\) (.*)');
+      final match = regExp.firstMatch(line);
+      if (match != null) {
+        final matchedText = match.group(1)!;
+        log(matchedText);
+      }
```
Comment (piotruela) [inline packages/patrol_log/lib/src/patrol_log_reader.dart:99]: Same for this one and the one below

File: packages/patrol_log/lib/src/entry.dart [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+  @override
+  String toString() => throw UnimplementedError('toString is not implemented');
+}
+
+enum EntryType {
+  step,
+  test,
+  log;
+}
```
Comment (piotruela) [inline packages/patrol_log/lib/src/entry.dart:24]: I guess that won't be necessary with sealed classes approach

File: packages/patrol_log/analysis_options.yaml [hunk 4]
```diff
@@ -0,0 +1 @@
+include: package:leancode_lint/analysis_options.yaml
```
Comment (piotruela) [inline packages/patrol_log/analysis_options.yaml:1]: Add newline at the end of the file

File: packages/patrol_finders/lib/src/custom_finders/patrol_tester.dart [hunk 5]
```diff
@@ -116,17 +117,60 @@ const defaultScrollMaxIteration = 15;
...
+  PatrolTester({
     required this.tester,
     required this.config,
   });
 
   /// Global configuration of this tester.
   final PatrolTesterConfig config;
 
+  /// Logs a message to the patrol log.
+  final PatrolLogWriter _patrolLog = PatrolLogWriter();
```
Comment (piotruela) [inline packages/patrol_finders/lib/src/custom_finders/patrol_tester.dart:129]: Can this be initialised in initialiser list of the class?

File: packages/patrol_finders/lib/src/custom_finders/patrol_tester.dart [hunk 6]
```diff
@@ -116,17 +117,60 @@ const defaultScrollMaxIteration = 15;
...
   /// Flutter's widget tester that this [PatrolTester] wraps.
   final WidgetTester tester;
 
+  /// Wraps a function with a log entry for the start and end of the function.
+  Future<T> wrapWithPatrolLog<T>({
+    required String action,
+    String? value,
+    Finder? finder,
+    required String color,
+    required Function function,
```
Comment (piotruela) [inline packages/patrol_finders/lib/src/custom_finders/patrol_tester.dart:140]: If you declare this paramteres as:
```suggestion
    required Future<T> Function() function,
```

Comment (piotruela) [reply packages/patrol_finders/lib/src/custom_finders/patrol_tester.dart:140]: Then you should be able to call the methods inside wrapper without casting:
`      return function();`

File: packages/patrol_finders/lib/src/custom_finders/patrol_tester.dart [hunk 7]
```diff
@@ -243,18 +287,26 @@ class PatrolTester {
...
-      await tester.tap(resolvedFinder.first);
-      await _performPump(
-        settlePolicy: settlePolicy,
-        settleTimeout: settleTimeout,
-      );
-    });
+    return wrapWithPatrolLog(
+      action: 'tap',
+      finder: finder,
+      color: '33m',
```
Comment (piotruela) [inline packages/patrol_finders/lib/src/custom_finders/patrol_tester.dart:295]: What color is this? I think this should also be wrapped with something more descriptive

File: packages/patrol_cli/lib/src/runner/patrol_command.dart [hunk 8]
```diff
@@ -181,6 +181,20 @@ abstract class PatrolCommand extends Command<int> {
...
+      'show-flutter-logs',
+      help: 'Show Flutter logs while running the tests.',
+    );
+  }
+
+  void usesShowTestSteps() {
+    argParser.addFlag(
+      'show-test-steps',
+      help: 'Show test steps while running the tests.',
+    );
```
Comment (piotruela) [inline packages/patrol_cli/lib/src/runner/patrol_command.dart:195]: I think we should consider setting this one to `true` by default. Default behaviour on Android doesn't give any information about passed/failed test.

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart [hunk 9]
```diff
@@ -152,15 +153,38 @@ class IOSTestBackend {
     IOSAppOptions options,
     Device device, {
     bool interruptible = false,
+    required bool showFlutterLogs,
+    required bool showTestSteps,
   }) async {
     await _disposeScope.run((scope) async {
-      final subject = '${options.description} on ${device.description}';
-      final task = _logger.task('Running $subject');
+      // Read patrol logs from logcat
```
Comment (piotruela) [inline packages/patrol_cli/lib/src/ios/ios_test_backend.dart:160]: Is it from logcat?

File: packages/patrol_cli/lib/src/commands/test.dart [hunk 10]
```diff
@@ -339,21 +343,34 @@ See https://github.com/leancodepl/patrol/issues/1316 to learn more.
...
 
     switch (device.targetPlatform) {
       case TargetPlatform.android:
-        action = () => _androidTestBackend.execute(android, device);
+        action = () => _androidTestBackend.execute(
+              android,
+              device,
+              showFlutterLogs: showFlutterLogs,
+              showTestSteps: showTestSteps,
+              flavor: flutterOpts.flavor,
```
Comment (piotruela) [inline packages/patrol_cli/lib/src/commands/test.dart:359]: Is that intended?

File: packages/patrol_cli/lib/src/commands/develop.dart [hunk 11]
```diff
@@ -327,8 +333,14 @@ class DevelopCommand extends PatrolCommand {
...
         appId = android.packageName;
-        action = () =>
-            _androidTestBackend.execute(android, device, interruptible: true);
+        action = () => _androidTestBackend.execute(
+              android,
+              device,
+              interruptible: true,
+              showFlutterLogs: showFlutterLogs,
+              showTestSteps: showTestSteps,
+              flavor: flutterOpts.flavor,
```
Comment (piotruela) [inline packages/patrol_cli/lib/src/commands/develop.dart:342]: Is that intended? Seems like unrelated with the PR code

File: packages/patrol_cli/lib/src/android/android_test_backend.dart [hunk 12]
```diff
@@ -186,9 +188,44 @@ class AndroidTestBackend {
...
+        runInShell: true,
+      )
+        ..disposedBy(scope);
+
+      var flavorPath = '';
+      if (flavor != null) {
+        flavorPath = 'flavors/$flavor/';
+      }
+      final reportPath =
+          'file://${_rootDirectory.childDirectory('android').path}/build/app/reports/androidTests/connected/${flavorPath}index.html';
```
Comment (piotruela) [inline packages/patrol_cli/lib/src/android/android_test_backend.dart:216]: This one probably won't work on Windows (due to different separated used there).

File: packages/patrol_cli/lib/src/android/android_test_backend.dart [hunk 13]
```diff
@@ -186,9 +188,44 @@ class AndroidTestBackend {
...
+    String? flavor,
     bool interruptible = false,
+    required bool showFlutterLogs,
+    required bool showTestSteps,
   }) async {
     await _disposeScope.run((scope) async {
+      // Read patrol logs from logcat
+      final processLogcat = await _processManager.start(
+        [
+          'adb',
```
Comment (piotruela) [inline packages/patrol_cli/lib/src/android/android_test_backend.dart:200]: We should use `adb` commands using a wrapper. It's already available in the scope of this class

File: packages/patrol/lib/src/native/native_automator2.dart [hunk 14]
```diff
@@ -96,19 +98,52 @@ class NativeAutomator2 {
...
   }
 
-  Future<T> _wrapRequest<T>(String name, Future<T> Function() request) async {
+  Future<T> _wrapRequest<T>(
+    String name,
+    Future<T> Function() request, {
+    bool enablePatrolLog = true,
+  }) async {
     _config.logger('$name() started');
+    final text = '\u001b[38;5;87m$name\u001b[0m \u001b[30m(native)\u001b[0m';
```
Comment (piotruela) [inline packages/patrol/lib/src/native/native_automator2.dart:107]: Enigmatic characters

File: packages/patrol/lib/src/native/native_automator.dart [hunk 15]
```diff
@@ -214,19 +216,52 @@ class NativeAutomator {
...
   }
 
-  Future<T> _wrapRequest<T>(String name, Future<T> Function() request) async {
+  Future<T> _wrapRequest<T>(
+    String name,
+    Future<T> Function() request, {
+    bool enablePatrolLog = true,
+  }) async {
     _config.logger('$name() started');
+    final text = '\u001b[38;5;87m$name\u001b[0m \u001b[30m(native)\u001b[0m';
```
Comment (piotruela) [inline packages/patrol/lib/src/native/native_automator.dart:225]: ditto

File: packages/patrol/lib/src/custom_finders/patrol_integration_tester.dart [hunk 16]
```diff
@@ -1,17 +1,22 @@
...
 class PatrolIntegrationTester extends finders.PatrolTester {
   /// Creates a new [PatrolIntegrationTester] which wraps [tester].
-  const PatrolIntegrationTester({
+  PatrolIntegrationTester({
     required super.tester,
     required super.config,
     required this.nativeAutomator,
     required this.nativeAutomator2,
-  });
+    required PatrolLogWriter patrolLog,
```
Comment (piotruela) [inline packages/patrol/lib/src/custom_finders/patrol_integration_tester.dart:15]: This one will sort of break the API (look here to [semver check](https://github.com/leancodepl/patrol/actions/runs/11615023075/job/32344576266?pr=2387). I know most of the users are not using this class directly but some of the projects do. Can we somehow leave this `PatrolLogWriter` as an optional parameter? At least until we introduce patrol 4.0

File: packages/patrol_log/lib/src/patrol_log_reader.dart [hunk 17]
```diff
@@ -0,0 +1,207 @@
...
+          (entry.status == TestEntryStatus.skip ||
+              entry.status == TestEntryStatus.start)) {
+        _singleEntries.add(PatrolSingleTestEntry([entry]));
+      } else if (entry is TestEntry) {
+        final lastEntry = _singleEntries.last;
+        lastEntry.entries.add(entry);
+      } else {
+        final lastEntry = _singleEntries.last;
+        lastEntry.entries.add(entry);
+      }
```
[the remark this answers is not in the document]
Comment (pdenert) [reply packages/patrol_log/lib/src/patrol_log_reader.dart:141]: It is not. This is some residue from changes during development. Deleted

File: packages/patrol_log/lib/src/step_entry.dart [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+        return '  $number.';
+      } else if (number < 100) {
+        return ' $number.';
+      } else {
+        return '$number.';
+      }
+    }
+
+    return '';
+  }
```
[the remark this answers is not in the document]
Comment (pdenert) [reply packages/patrol_log/lib/src/step_entry.dart:56]: It's a step number in single test. We're adding spaces to keep proper indentation.
Zrzut ekranu 2024-11-4 o 23 54 03

[the remark this answers is not in the document]
Comment (piotruela) [reply packages/patrol_log/lib/src/step_entry.dart:56]: Makes sense. I would rename it to `index` and add some comment to explain what is happening and why

File: packages/patrol_log/lib/src/entry.dart (hunk 3 above)
Comment (pdenert) [reply packages/patrol_log/lib/src/entry.dart:24]: We need to somehow pass type of the step to json, and use this info in `parseEntry` in patrol_log_reader so this was my way. Do you have better idea?

File: packages/patrol/lib/src/custom_finders/patrol_integration_tester.dart (hunk 16 above)
Comment (pdenert) [reply packages/patrol/lib/src/custom_finders/patrol_integration_tester.dart:15]: I've change it, so it won't be passed as argument, but initialized locally

File: packages/patrol_cli/lib/src/ios/ios_test_backend.dart (hunk 9 above)
Comment (pdenert) [reply packages/patrol_cli/lib/src/ios/ios_test_backend.dart]: It is not. Changed

File: packages/patrol_cli/lib/src/runner/patrol_command.dart (hunk 8 above)
Comment (pdenert) [reply packages/patrol_cli/lib/src/runner/patrol_command.dart:195]: This flag steer only showing test steps, the test status is always displayed. But as we discussed internally, I will change this flag to hide-test-steps

File: packages/patrol_log/lib/src/entry.dart (hunk 3 above)
Comment (piotruela) [reply packages/patrol_log/lib/src/entry.dart:24]: You're right, we still need that enum

File: packages/patrol_cli/lib/src/commands/test.dart (hunk 10 above)
Comment (pdenert) [reply packages/patrol_cli/lib/src/commands/test.dart:359]: Yes, I need this to generate proper android report path

File: packages/patrol_cli/lib/src/commands/develop.dart (hunk 11 above)
Comment (pdenert) [reply packages/patrol_cli/lib/src/commands/develop.dart:342]: As above

File: packages/patrol_log/lib/src/patrol_log_reader.dart [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+    switch (type) {
+      case EntryType.step:
+        return StepEntry.fromJson(json);
+      case EntryType.test:
+        return TestEntry.fromJson(json);
+      case EntryType.log:
+        return LogEntry.fromJson(json);
+      default:
+        throw UnimplementedError('Unknown entry type');
+    }
```
Comment (piotruela) [inline packages/patrol_log/lib/src/patrol_log_reader.dart:151]: Could be switch expression

File: packages/patrol_log/lib/src/entry.dart [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+    switch (name) {
+      case 'step':
+        return EntryType.step;
+      case 'test':
+        return EntryType.test;
+      case 'log':
+        return EntryType.log;
+      default:
+        throw ArgumentError('Unknown EntryType: $name');
+    }
```
Comment (piotruela) [inline packages/patrol_log/lib/src/entry.dart:51]: Could be switch expression

File: packages/patrol_log/lib/src/patrol_log_reader.dart [hunk 21]
```diff
@@ -0,0 +1,295 @@
...
+    //       log(entry.pretty(number: stepsCounter));
+    //     }
+    //   } else if (entry is LogEntry) {
+    //     _singleEntries.last.addEntry(entry);
+    //     logsCounter++;
+
+    //     // Print the log entry to the console.
+    //     log(entry.pretty());
+    //   }
+    // });
```
Comment (piotruela) [inline packages/patrol_log/lib/src/patrol_log_reader.dart:256]: Can be deleted probably

File: packages/patrol_log/lib/src/patrol_single_test_entry.dart [hunk 22]
```diff
@@ -0,0 +1,42 @@
...
+  /// The status of the test.
+  TestEntryStatus get status =>
+      closingTestEntry?.status ?? openingTestEntry.status;
+
+  /// The name of the test with the path.
+  String get nameWithPath => openingTestEntry.nameWithPath;
+
+  /// The execution time of the test.
+  Duration get executionTime =>
+      closingTestEntry!.executionTime(openingTestEntry.timestamp);
```
Comment (piotruela) [inline packages/patrol_log/lib/src/patrol_single_test_entry.dart:41]: Can we somehow get rid of the `!` from here? Maybe we can live with `executionTime` being nullable getter?

File: packages/patrol_log/lib/src/step_entry.dart [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+      if (number < 10) {
+        return '  $number.';
+      } else if (number < 100) {
+        return ' $number.';
+      } else {
+        return '$number.';
+      }
+    }
+
+    return '';
```
Comment (piotruela) [inline packages/patrol_log/lib/src/step_entry.dart:45]:
```suggestion
    return switch (number) {
      == null => '',
      < 10 => '  $number.',
      < 100 => ' $number.',
      _ => '$number.',
    };
```

**Review by piotruela — COMMENTED** [review]
Looking good. Just a few nitpick
